// @polyjuice-provider/base/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigningMessageType = exports.ShortAddressType = void 0;
var ShortAddressType;
(function (ShortAddressType) {
    ShortAddressType[ShortAddressType["eoaAddress"] = 0] = "eoaAddress";
    ShortAddressType[ShortAddressType["contractAddress"] = 1] = "contractAddress";
    ShortAddressType[ShortAddressType["notExistEoaAddress"] = 2] = "notExistEoaAddress";
    ShortAddressType[ShortAddressType["notExistContractAddress"] = 3] = "notExistContractAddress";
    ShortAddressType[ShortAddressType["creatorAddress"] = 4] = "creatorAddress";
})(ShortAddressType = exports.ShortAddressType || (exports.ShortAddressType = {}));
var SigningMessageType;
(function (SigningMessageType) {
    SigningMessageType[SigningMessageType["withPrefix"] = 0] = "withPrefix";
    SigningMessageType[SigningMessageType["noPrefix"] = 1] = "noPrefix";
})(SigningMessageType = exports.SigningMessageType || (exports.SigningMessageType = {}));
